samtranslator/model/sam_resources.py

```diff
@@ -2407,3 +2418,148 @@ def _set_resolver_code_settings(resolver_code_settings: aws_serverless_graphqlap
+
+    def _construct_appsync_function_configurations(
+        self,
+        functions: Dict[str, aws_serverless_graphqlapi.Function],
+        datasources: Optional[Dict[str, aws_serverless_graphqlapi.DynamoDBDataSource]],
+        api_id: Intrinsicable[str],
+    ) -> List[Resource]:
+        resources: List[Resource] = []
+
+        none_datasource = self._check_and_construct_none_datasource(functions, api_id)
+
+        if none_datasource:
+            resources.append(none_datasource)
+
+        for relative_id, function in functions.items():
+            func_config = FunctionConfiguration(
+                logical_id=self.logical_id + relative_id,
+                depends_on=self.depends_on,
+                attributes=self.resource_attributes,
+            )
+
+            # "Id" is a mutually exclusive property to every other property. If this function has it
+            # defined, then we make sure it's the only property, and continue to next function.
+            if function.Id:
+                keys = remove_none_items(function.dict()).keys()  # remove undefined properties, then get keys
+                if len(keys) != 1:
+                    raise InvalidResourceException(
+                        relative_id, "'Id' cannot be defined with other properties in Function."
+                    )
+                continue
+
+            if not function.InlineCode and not function.CodeUri:
+                raise InvalidResourceException(relative_id, "One of 'InlineCode' or 'CodeUri' must be set.")
+
+            if function.InlineCode and function.CodeUri:
+                raise InvalidResourceException(
+                    relative_id, "Both 'InlineCode' and 'CodeUri' cannot be defined at the same time."
+                )
+
+            if function.InlineCode and not function.Runtime:
+                raise InvalidResourceException(
+                    relative_id,
+                    "'Runtime' must be defined when 'InlineCode' is defined.",
+                )
+
+            if not function.DataSource and not function.DataSourceName:
+                raise InvalidResourceException(relative_id, "One of 'DataSource' or 'DataSourceName' must be set.")
+
+            if function.DataSource and function.DataSourceName:
+                raise InvalidResourceException(
+                    relative_id, "Both 'DataSourceType' and 'DataSourceName' cannot be defined at the same time."
+                )
+
+            func_config.Code = function.InlineCode
+            func_config.CodeS3Location = function.CodeUri
+            func_config.ApiId = api_id
+            func_config.Name = function.Name or relative_id
+            func_config.DataSourceName = self._parse_datasource_name(relative_id, function, datasources)
+            func_config.MaxBatchSize = cast(PassThrough, function.MaxBatchSize)
+            func_config.Description = cast(PassThrough, function.Description)
```

Comment:
I think we discussed this before, `PassThroughProp` does not type it as `Any` so we need to cast for `PassThrough` types

Reply:
Oh right, because we override the property getter and return `__root__` (of type `Any`), but mypy thinks it returns `PassThroughProp`...

Reply:
Since it's used so often and the intention/rationale isn't obvious from code, I'm tempted for us to just use `function.Description.__root__` to access the value to avoid the cast altogheter... It's ugly but it's not clear currently either. Or maybe a tiny documented wrapper function.

Reply:
Just to update this conversation so it's noted, we are using a wrapper function for `PassThroughProp` type values now to avoid confusion around the cast.